src/core/inductive.ml

```diff
@@ -0,0 +1,213 @@
+    [codom [x1;...;xn]]. *)
+let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> tbox =
+  fun pos ind_sym codom ->
+  let rec aux : tvar list -> term -> tbox = fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> codom (List.rev_map _Vari xs)
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
+  in aux [] !(ind_sym.sym_type)
+
+(** [gen_rec_type ss pos ind_sym cons_list] generates the induction principle
+    of the inductive type [ind_sym] (and its position [pos]) with the cons-
+    tructors [cons_list]. Each recursive argument is followed by its induction
+    hypothesis. For instance, with [inductive T:TYPE := c: T->T->T], we have
+    [ind_T: Πp:T->Prop, (Πx1:T, π(p x1)-> Πx2:T, π(p x2)-> π(p(c x1 x2)) ->
+    Πx:T, π(p x)]. In general, thanks to this function, the command
+    inductive T : Π(x1:A1),...,Π(xm:Am), TYPE := c1:T1 | ... | cn:Tn generates
+    ind_T: Π(p:Π(x1:A1),...,Π(xm:Am), T x1 ... xm -> Prop), U1 -> ... -> Un ->
+    (Π(x1:A1),...,Π(xm:Am), Π(x:T x1 ... xm), π(p x1 ... xm x))
+    where Ui is the clause associated to the constructor ci. *)
+let gen_rec_type : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos ind_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let c = get_config ss pos in
+  let p = Bindlib.new_var mkfree "p" in
+  let app  : tbox -> tbox list -> tbox = List.fold_left _Appl in
+  let fapp : sym  -> tbox list -> tbox = fun f ts -> app (_Symb f) ts in
+  let prf_of_p : tbox list -> tbox -> tbox = fun ts t ->
+    fapp c.symb_prf [_Appl (app (_Vari p) ts) t]
+  in
+
+  (* STEP 1: Create the type of the property p *)
+  let prop = _Symb c.symb_Prop in
+  let codom ts = _Impl (fapp ind_sym ts) prop in
+  let p_type = gen_ind_typ_codom pos ind_sym codom in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of cons_sym] creates a clause according to a constructor
+     [cons_sym]. *)
+  let case_of : sym -> tbox = fun cons_sym ->
+    let codom : term list -> tvar list -> tbox -> tbox = fun ts xs _ ->
+      prf_of_p (List.map lift ts)
+        (app (_Symb cons_sym) (List.rev_map _Vari xs))
+    in
+    let get_name : tbinder -> int -> tvar * term = fun b _ ->
+      Basics.unbind_name b
+    in
+    let build_rec_hyp : term list -> tvar -> tbox = fun ts x ->
+      prf_of_p (List.map lift ts) (_Vari x)
+    in
+    let domrec : term -> tvar -> tbox -> tbox -> tbox =
+      fun a x rec_hyp next ->
+      _Prod (lift a) (Bindlib.bind_var x (_Impl rec_hyp next))
+    in
+    let dom : term -> tvar -> tbox -> tbox = fun a x next ->
+      _Prod (lift a) (Bindlib.bind_var x next)
+    in
+    let f_rec : tbox -> tvar -> tbox -> tbox = fun a _ _ -> a in
+    let f_not_rec : tbox -> tvar -> tbox = fun a _ -> a in
+    let acc : tbox = _Type in
```

Comment:
acc can be unfolded here

Reply:
ignore this comment: I didn't acc was among the arguments

Reply:
I don't understand.